### ISSUE TITLE
Anomaly "grounding a non evar-free term"

### ISSUE THREAD

**Issue by ppedrot** [coq/coq#7538]
Yet another example of the anomaly. Not fixed by #7449.

```
CoInductive Stream (A : Type) : Type :=
    Cons { hd : A; tl : Stream A }.

Fixpoint Str_nth_tl {A : Type} (n:nat) (s:Stream A) : Stream A :=
  match n with
  | O => s
  | S m => Str_nth_tl _ m (tl _ s)
  end.
```

**Comment by uchwings** [comment]
As of a git checkout from a few days ago, the following trivial implementation of fold over a list also gives this error:
```
Require Import List.

Fixpoint listFold f list :=
  match list with
    | nil => nil
    | x :: rest => f x :: listFold f rest
  end
.
```

**Comment by ejgallego** [comment]
I can provide an ad-hoc patch for this one, however it'd be good if we could merge the proposed solution in #7530

**Comment by ejgallego** [comment]
This was fixed by #8834
